SimG4CMS/Forward/src/BscNumberingScheme.cc

```diff
@@ -13,69 +13,51 @@
 
-int BscNumberingScheme::detectorLevel(const G4Step* aStep) const {
-  //Find number of levels
-  const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
-  return (touch) ? ((touch->GetHistoryDepth()) + 1) : 0;
-}
-
-void BscNumberingScheme::detectorLevel(const G4Step* aStep, int& level, int* copyno, G4String* name) const {
-  //Get name and copy numbers
-  if (level > 0) {
-    const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
-    for (int ii = 0; ii < level; ++ii) {
-      int i = level - ii - 1;
-      name[ii] = ForwardName::getName(touch->GetVolume(i)->GetName());
-      copyno[ii] = touch->GetReplicaNumber(i);
-    }
-  }
-}
-
 unsigned int BscNumberingScheme::getUnitID(const G4Step* aStep) const {
   unsigned int intindex = 0;
-  int level = detectorLevel(aStep);
+
+  //Find number of levels
+  const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
+  int level = (touch) ? ((touch->GetHistoryDepth()) + 1) : 0;
 
   LogDebug("BscSim") << "BscNumberingScheme number of levels= " << level;
 
   if (level > 0) {
-    int* copyno = new int[level];
-    G4String* name = new G4String[level];
-    detectorLevel(aStep, level, copyno, name);
-
     int det = 0;
     int zside = 0;
     int station = 0;
     for (int ich = 0; ich < level; ich++) {
+      int copyno = touch->GetReplicaNumber(level - ich - 1);
+      G4String name = ForwardName::getName(touch->GetVolume(level - ich - 1)->GetName());
       // new and old set up configurations are possible:
-      if (name[ich] == "BSC1" || name[ich] == "BSC2") {
-        zside = copyno[ich] - 1;
-      } else if (name[ich] == "BSCTrap") {
+      if (name == "BSC1" || name == "BSC2") {
+        zside = copyno - 1;
+      } else if (name == "BSCTrap") {
         det = 0;
-        station = 2 * (copyno[ich] - 1);
-      } else if (name[ich] == "BSCTubs") {
+        station = 2 * (copyno - 1);
+      } else if (name == "BSCTubs") {
         det = 1;
-        station = copyno[ich] - 1;
-      } else if (name[ich] == "BSCTTop") {
+        station = copyno - 1;
+      } else if (name == "BSCTTop") {
         ++station;
-      } else if (name[ich] == "BSC2Pad") {
+      } else if (name == "BSC2Pad") {
         det = 2;
-        station = copyno[ich] - 1;
+        station = copyno - 1;
       }
 
       LogDebug("BscSim") << "BscNumberingScheme  "
-                         << "ich=" << ich << "copyno" << copyno[ich] << "name=" << name[ich];
+                         << "ich=" << ich << "copyno" << copyno << "name=" << name;
     }
     intindex = packBscIndex(zside, det, station);
     LogDebug("BscSim") << "BscNumberingScheme : det " << det << " zside " << zside << " station " << station
                        << " UnitID 0x" << std::hex << intindex << std::dec;
 
-    for (int ich = 0; ich < level; ich++)
-      LogDebug("BscSim") << " name = " << name[ich] << " copy = " << copyno[ich];
-
+    for (int ich = 0; ich < level; ich++) {
```

Comment:
Why we need this loop? LogDebug may be moved inside previous loop after line 31

Reply:
Will do